akka-actor-typed/src/main/scala/akka/actor/typed/Behavior.scala

```diff
@@ -162,17 +157,32 @@ object Behavior {
    * and then the resulting behavior is returned.
    */
   def start[T](behavior: Behavior[T], ctx: TypedActorContext[T]): Behavior[T] = {
-    // note that this can't be @tailrec, but normal stack of interceptors and similar shouldn't be
-    // that deep, and if they are it's most likely a bug which will be seen as StackOverflowError
-    behavior match {
-      case innerDeferred: DeferredBehavior[T]          => start(innerDeferred(ctx), ctx)
-      case wrapped: InterceptorImpl[T, Any] @unchecked =>
-        // make sure that a deferred behavior wrapped inside some other behavior is also started
-        val startedInner = start(wrapped.nestedBehavior, ctx.asInstanceOf[TypedActorContext[Any]])
-        if (startedInner eq wrapped.nestedBehavior) wrapped
-        else wrapped.replaceNested(startedInner)
-      case _ => behavior
+    @tailrec
+    def internalStart(
+        behavior: Behavior[T],
+        ctx: TypedActorContext[T],
+        interceptorsStack: List[InterceptorImpl[T, Any]] = Nil): Behavior[T] = {
+      // normal stack of interceptors and similar shouldn't be that deep,
+      // and if they are , which will generate a large interceptorsStack,
+      // it's most likely a bug which maybe be seen as OOM.
+      behavior match {
+        case innerDeferred: DeferredBehavior[T]          => internalStart(innerDeferred(ctx), ctx)
+        case wrapped: InterceptorImpl[T, Any] @unchecked =>
+          // make sure that a deferred behavior wrapped inside some other behavior is also started
+          val nestedBehavior = wrapped.nestedBehavior
+          internalStart(nestedBehavior.narrow, ctx.asInstanceOf[TypedActorContext[T]], wrapped :: interceptorsStack)
+        case _ =>
+          interceptorsStack match {
+            case Nil => behavior
+            case _ =>
+              interceptorsStack.foldLeft(behavior) {
+                case (b, acc) => acc.replaceNested(b.unsafeCast)
+              }
+          }
+
+      }
     }
+    internalStart(behavior, ctx)
```

Comment:
Isn't this similar to the attempt in https://github.com/akka/akka/pull/27364 ?
It's replacing the stack with a collection. I'm not sure there are any advantages of doing that.

Reply:
I have forgotten that. I came up with this once I read in detail of the Behavior code. With this change, it will work more concisely, at least it doesn't harm, isn't it? Different deploy environments   XSS configs various too.

Reply:
I ran a temporary benchmark comparing the two and this is about 30% slower.

BEFORE:
```
[info] Benchmark                                               (batchSize)    (dispatcher)                                         (mailbox)  (tpt)   Mode  Cnt        Score        Error   Units
[info] TypedActorBenchmark.start                                        50  fjp-dispatcher  akka.dispatch.SingleConsumerOnlyUnboundedMailbox     50  thrpt    6  8048211,084 ± 156294,353   ops/s
```

AFTER:
```
[info] Benchmark                                               (batchSize)    (dispatcher)                                         (mailbox)  (tpt)   Mode  Cnt        Score        Error   Units
[info] TypedActorBenchmark.start                                        50  fjp-dispatcher  akka.dispatch.SingleConsumerOnlyUnboundedMailbox     50  thrpt    6  5772894,593 ± 257261,651   ops/s
```

Used the following in TypedActorBenchmark:
```scala
  private val behv: Behavior[Start] =
    Behaviors
      .supervise[Start] {
        Behaviors.setup[Start] { _ =>
          Behaviors.withMdc(Map.empty[String, String]) {
            Behaviors.receiveMessage[Start](_ => Behaviors.same)
          }
        }
      }
      .onFailure[Exception](SupervisorStrategy.restart)

  def startSupervisor(iteration: Int): Behavior[Start] =
    Behaviors.receive { (ctx, msg) =>
      msg match {
        case Start(respondTo) =>
          val startNanoTime = System.nanoTime()

          var i = 0
          while (i < iteration) {
            Behavior.start(behv, ctx)
            i += 1
          }

          respondTo ! Completed(startNanoTime)
          Behaviors.same
      }
    }

  @Benchmark
  @OperationsPerInvocation(iterations)
  def start(): Unit = {
    Await.result(system.ask(Start), timeout)
  }
```

Reply:
Then it's not worth it.